mlflow/projects/__init__.py

```diff
@@ -189,6 +194,29 @@ def _is_local_uri(uri):
     return not _GIT_URI_REGEX.match(uri)
 
 
+def _is_valid_version(repo, version):
+    """Verify the commit version of branch name is valid, raise exception if otherwise."""
+    from git.exc import GitCommandError
+    try:
+        return repo.git.rev_parse("--verify", version)
+    except GitCommandError:
+        # Second attempt to see if this is a branch
+        try:
+            return repo.git.rev_parse("--verify", "refs/remotes/origin/%s" % version)
+        except GitCommandError:
+            raise ExecutionException("'%s' is invalid or does not exist in the Git project." %
+                                     version)
+
+
+def _is_valid_branch_name(uri, branch):
+    """Returns True if the branch exists in the Git project."""
+    if branch is not None:
+        parsed_uri = _parse_subdirectory(uri)[0]
+        from git import cmd
+        return cmd.Git().ls_remote("--heads", parsed_uri, branch)
```

Comment:
`_fetch_git_repo()` would have failed without network access, right? :) I was hesitate on using `rev_parse` since the `work_dir` returned from `_fetch_git_repo()` already has the git project sub directory appended. But it sounds better avoiding another remote lookup so I have made the code change.